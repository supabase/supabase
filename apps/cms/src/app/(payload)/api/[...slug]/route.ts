/* THIS FILE WAS GENERATED AUTOMATICALLY BY PAYLOAD. */
/* DO NOT MODIFY IT BECAUSE IT COULD BE REWRITTEN AT ANY TIME. */
import config from '@payload-config'
import '@payloadcms/next/css'
import {
  REST_DELETE,
  REST_GET,
  REST_OPTIONS,
  REST_PATCH,
  REST_POST,
  REST_PUT,
} from '@payloadcms/next/routes'

export const GET = REST_GET(config)
export const POST = REST_POST(config)
export const DELETE = REST_DELETE(config)
export const PATCH = REST_PATCH(config)
export const PUT = REST_PUT(config)
export const OPTIONS = REST_OPTIONS(config)
