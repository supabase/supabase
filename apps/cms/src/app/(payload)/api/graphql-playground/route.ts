/* THIS FILE WAS GENERATED AUTOMATICALLY BY PAYLOAD. */
/* DO NOT MODIFY IT BECAUSE IT COULD BE REWRITTEN AT ANY TIME. */
import config from '@payload-config'
import '@payloadcms/next/css'
import { GRAPHQL_PLAYGROUND_GET } from '@payloadcms/next/routes'

export const GET = GRAPHQL_PLAYGROUND_GET(config)
