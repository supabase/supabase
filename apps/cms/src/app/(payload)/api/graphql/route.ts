/* THIS FILE WAS GENERATED AUTOMATICALLY BY PAYLOAD. */
/* DO NOT MODIFY IT BECAUSE IT COULD BE REWRITTEN AT ANY TIME. */
import config from '@payload-config'
import { GRAPHQL_POST, REST_OPTIONS } from '@payloadcms/next/routes'

export const POST = GRAPHQL_POST(config)

export const OPTIONS = REST_OPTIONS(config)
