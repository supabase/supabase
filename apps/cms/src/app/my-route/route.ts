import configPromise from '@payload-config'
import { getPayload } from 'payload'

export const GET = async () => {
  const payload = await getPayload({
    config: configPromise,
  })

  const data = await payload.find({
    collection: 'users',
  })

  return Response.json(data)
}
