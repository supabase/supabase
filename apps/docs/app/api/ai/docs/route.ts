import { SupabaseClient } from '@supabase/supabase-js'
import { ApplicationError, UserError, clippy } from 'ai-commands/edge'
import { NextRequest, NextResponse } from 'next/server'
import OpenAI from 'openai'

export const runtime = 'edge'
/* To avoid OpenAI errors, restrict to the Vercel Edge Function regions that
  overlap with the OpenAI API regions.

  Reference for Vercel regions: https://vercel.com/docs/edge-network/regions#region-list
  Reference for OpenAI regions: https://help.openai.com/en/articles/5347006-openai-api-supported-countries-and-territories
  */
export const preferredRegion = [
  'arn1',
  'bom1',
  'cdg1',
  'cle1',
  'cpt1',
  'dub1',
  'fra1',
  'gru1',
  'hnd1',
  'iad1',
  'icn1',
  'kix1',
  'lhr1',
  'pdx1',
  'sfo1',
  'sin1',
  'syd1',
]

const openAiKey = process.env.OPENAI_API_KEY
const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL as string
const supabaseServiceKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY as string

export async function POST(req: NextRequest) {
  if (!openAiKey || !supabaseUrl || !supabaseServiceKey) {
    return NextResponse.json(
      { error: 'Missing environment variables for AI features.' },
      { status: 500 }
    )
  }

  const openai = new OpenAI({ apiKey: openAiKey })
  const supabaseClient = new SupabaseClient(supabaseUrl, supabaseServiceKey)

  try {
    const { messages } = (await req.json()) as {
      messages: { content: string; role: 'user' | 'assistant' }[]
    }

    if (!messages) {
      throw new UserError('Missing messages in request data')
    }

    const response = await clippy(openai, supabaseClient, messages)

    // Proxy the streamed SSE response from OpenAI
    return new NextResponse(response.body, {
      headers: {
        'Content-Type': 'text/event-stream',
      },
    })
  } catch (error: unknown) {
    console.error(error)
    if (error instanceof UserError) {
      return NextResponse.json({ error: error.message, data: error.data }, { status: 400 })
    } else if (error instanceof ApplicationError) {
      console.error(`${error.message}: ${JSON.stringify(error.data)}`)
    } else {
      console.error(error)
    }

    return NextResponse.json(
      { error: 'There was an error processing your request' },
      { status: 500 }
    )
  }
}
