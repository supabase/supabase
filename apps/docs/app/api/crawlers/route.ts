import { toHtml } from 'hast-util-to-html'
import { fromMarkdown } from 'mdast-util-from-markdown'
import { mdxFromMarkdown } from 'mdast-util-mdx'
import { toHast } from 'mdast-util-to-hast'
import { mdxjs } from 'micromark-extension-mdxjs'
import { redirect } from 'next/navigation'
import { visit } from 'unist-util-visit'

import { REFERENCES } from '~/content/navigation.references'
import {
  getFlattenedSections,
  getFunctionsList,
  getTypeSpec,
} from '~/features/docs/Reference.generated.singleton'
import { getRefMarkdown } from '~/features/docs/Reference.mdx'
import type { MethodTypes } from '~/features/docs/Reference.typeSpec'
import type { AbbrevApiReferenceSection } from '~/features/docs/Reference.utils'
import { notFoundLink } from '~/features/recommendations/NotFound.utils'
import { BASE_PATH } from '~/lib/constants'

export async function GET(request: Request) {
  const url = new URL(request.url)
  let [, , lib, maybeVersion, slug] = url.pathname.split('/')

  const libraryMeta = REFERENCES[lib]

  const isVersion = /^v\d+$/.test(maybeVersion)
  const version = isVersion ? maybeVersion : libraryMeta.versions[0]
  if (!isVersion) {
    slug = maybeVersion
  }

  let section: AbbrevApiReferenceSection
  let sectionsWithUrl: Array<AbbrevApiReferenceSection & { url: URL }>
  try {
    const flattenedSections = (await getFlattenedSections(lib, version)) ?? []
    sectionsWithUrl = flattenedSections.map((section) => {
      const url = new URL(request.url)
      url.pathname = [BASE_PATH, 'reference', lib, isVersion ? version : null, section.slug]
        .filter(Boolean)
        .join('/')

      return {
        ...section,
        url,
      }
    })
    section = flattenedSections.find(
      (section) =>
        (section.type === 'markdown' || section.type === 'function') && section.slug === slug
    )
  } catch {}

  if (!section) {
    redirect(notFoundLink(`${lib}/${slug}`))
  }

  const html = htmlShell(
    lib,
    isVersion ? version : null,
    slug,
    section,
    libraryNav(sectionsWithUrl) + (await sectionDetails(lib, isVersion ? version : null, section))
  )
  const response = new Response(html)
  response.headers.set('Content-Type', 'text/html; charset=utf-8')

  return response
}

function htmlShell(
  lib: string,
  version: string | null,
  slug: string,
  section: AbbrevApiReferenceSection,
  body: string
) {
  const libraryName = REFERENCES[lib].name
  let title = libraryName + ': ' + section.title ?? ''

  return (
    '<!doctype html><html>' +
    '<head>' +
    `<title>${title} | Supabase Docs</title>` +
    `<meta name="description" content="Supabase API reference for ${libraryName}${section.title ? ': ' + section.title : ''}">` +
    `<meta name="og:image" content="https://supabase.com/docs/img/supabase-og-image.png">` +
    `<meta name="twitter:image" content="https://supabase.com/docs/img/supabase-og-image.png">` +
    `<link rel="canonical" href="https://supabase.com/docs/reference/${lib}` +
    (version ? '/' + version : '') +
    (slug ? '/' + slug : '') +
    `">` +
    '</head>' +
    '<body>' +
    body +
    '</body></html>'
  )
}

function libraryNav(sections: Array<AbbrevApiReferenceSection & { url: URL }>) {
  return (
    '<nav><ul>' +
    sections
      .map((section) => `<li><a href="${section.url}">${section.title ?? ''}</a></li>`)
      .join('') +
    '</ul></nav>'
  )
}

async function sectionDetails(lib: string, version: string, section: AbbrevApiReferenceSection) {
  const libraryName = REFERENCES[lib].name
  let result = '<h1>' + (libraryName + ': ' + section.title ?? '') + '</h1>'

  if (section.type === 'markdown') {
    result += await markdown(lib, version, section)
  } else {
    result += await functionDetails(lib, version, section)
  }

  return result
}

async function markdown(lib: string, version: string | null, section: AbbrevApiReferenceSection) {
  const dir = !!section.meta?.shared ? 'shared' : lib + (version ? '/' + version : '')

  let content = await getRefMarkdown(dir + '/' + section.slug)
  content = mdxToHtml(content)
  return content
}

async function functionDetails(
  lib: string,
  version: string | null,
  section: AbbrevApiReferenceSection
) {
  const libraryMeta = REFERENCES[lib]

  const fns = await getFunctionsList(lib, version ?? libraryMeta.versions[0])
  const fn = fns!.find((fn) => fn.id === section.id)
  if (!fn) return ''

  let types: MethodTypes | undefined
  if (libraryMeta.typeSpec && '$ref' in fn) {
    types = await getTypeSpec(fn['$ref'] as string)
  }

  const fullDescription = [
    types?.comment?.shortText,
    'description' in fn && (fn.description as string),
    'notes' in fn && (fn.notes as string),
  ]
    .filter((x) => typeof x === 'string')
    .map(mdxToHtml)
    .join('')

  const parameters = parametersToHtml(fn, types)
  const examples = examplesToHtml(fn)

  return fullDescription + parameters + examples
}

function mdxToHtml(markdownUnescaped: string): string {
  const markdown = markdownUnescaped.replace(/(?<!\\)\{/g, '\\{').replace(/(?<!\\)\}/g, '\\}')

  const mdast = fromMarkdown(markdown, {
    extensions: [mdxjs()],
    mdastExtensions: [mdxFromMarkdown()],
  })

  visit(mdast, 'text', (node) => {
    node.value = node.value.replace(/\n/g, ' ')
  })
  if (!mdast) return ''

  const hast = toHast(mdast)
  if (!hast) return ''

  // @ts-ignore
  const html = toHtml(hast)

  return html
}

function parametersToHtml(fn: any, types: MethodTypes | undefined) {
  let result = '<h2 id="parameters">Parameters</h2>'

  if ('overwriteParams' in fn || 'params' in fn) {
    const params = fn.overwriteParams ?? fn.params
    if (params.length === 0) return ''

    result +=
      '<ul>' +
      params
        .map(
          (param) =>
            '<li>' +
            `<h3>${param.name}</h3>` +
            `<span>${param.isOptional ? '(Optional)' : '(Required)'}</span>` +
            `<p>${param.description}</p>` +
            '</li>'
        )
        .join('') +
      '</ul>'

    return result
  }

  if (!types?.params || types.params.length === 0) return ''

  result +=
    '<ul>' +
    types.params
      .map(
        (param) =>
          '<li>' +
          `<h3>${String(param.name)}</h3>` +
          `<span>${param.isOptional ? '(Optional)' : '(Required)'}</span>` +
          `<p>${param.comment?.shortText ?? ''}</p>` +
          '</li>'
      )
      .join('') +
    '</ul>'

  return result
}

function examplesToHtml(fn: any) {
  if (!fn.examples || fn.examples.length === 0) return ''

  let result = '<h2 id="examples">Examples</h2>'

  result += fn.examples
    .map((example) => `<h3>${example.name ?? ''}</h3>` + mdxToHtml(example.code ?? ''))
    .join('')

  return result
}
