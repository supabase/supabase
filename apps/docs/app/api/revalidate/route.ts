import { createClient } from '@supabase/supabase-js'
import { revalidateTag } from 'next/cache'
import { headers } from 'next/headers'
import { type NextRequest } from 'next/server'
import { z } from 'zod'

import { type Database } from 'common'

enum AuthorizationLevel {
  Unauthorized,
  Basic,
  Override,
}

const requestBodySchema = z.object({
  tags: z.array(z.string()),
})

export const POST = handleError(_handleRevalidateRequest)

export async function _handleRevalidateRequest(request: NextRequest) {
  const requestHeaders = headers()
  const authorization = requestHeaders.get('Authorization')
  if (!authorization) {
    return new Response('Missing Authorization header', { status: 401 })
  }

  const basicKeys = process.env.DOCS_REVALIDATION_KEYS?.split(/\s*,\s*/) ?? []
  const overrideKeys = process.env.DOCS_REVALIDATION_OVERRIDE_KEYS?.split(/\s*,\s*/) ?? []
  if (basicKeys.length === 0 && overrideKeys.length === 0) {
    console.error('No keys configured for revalidation')
    return new Response('Internal server error', {
      status: 500,
    })
  }

  let authorizationLevel = AuthorizationLevel.Unauthorized

  const token = authorization.replace(/^Bearer\s+/, '')
  if (overrideKeys.includes(token)) {
    authorizationLevel = AuthorizationLevel.Override
  } else if (basicKeys.includes(token)) {
    authorizationLevel = AuthorizationLevel.Basic
  }

  if (authorizationLevel === AuthorizationLevel.Unauthorized) {
    return new Response('Invalid Authorization header', { status: 401 })
  }

  let result: z.infer<typeof requestBodySchema>
  try {
    result = requestBodySchema.parse(await request.json())
  } catch (error) {
    console.error(error)
    return new Response(
      'Malformed request body: should be a JSON object with a "tags" array of strings.',
      { status: 400 }
    )
  }

  const supabaseAdmin = createClient<Database>(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.SUPABASE_SECRET_KEY!
  )

  if (authorizationLevel === AuthorizationLevel.Basic) {
    const { data: lastRevalidation, error } = await supabaseAdmin.rpc(
      'get_last_revalidation_for_tags',
      {
        tags: result.tags,
      }
    )
    if (error) {
      console.error(error)
      return new Response('Internal server error', { status: 500 })
    }

    const sixHoursAgo = new Date()
    sixHoursAgo.setHours(sixHoursAgo.getHours() - 6)
    if (lastRevalidation?.some((revalidation) => new Date(revalidation.created_at) > sixHoursAgo)) {
      return new Response(
        'Your request includes a tag that has been revalidated within the last 6 hours. You can override this limit by authenticating with Override permissions.',
        {
          status: 429,
        }
      )
    }
  }

  const { error } = await supabaseAdmin
    .from('validation_history')
    .insert(result.tags.map((tag) => ({ tag })))
  if (error) {
    console.error('Failed to update revalidation table: %o', error)
  }

  result.tags.forEach((tag) => {
    revalidateTag(tag)
  })

  return new Response(null, {
    status: 204,
    headers: {
      'Cache-Control': 'no-cache',
    },
  })
}

function handleError(handleRequest: (request: NextRequest) => Promise<Response>) {
  return async function (request: NextRequest) {
    try {
      const response = await handleRequest(request)
      return response
    } catch (error) {
      console.error(error)
      return new Response('Internal server error', { status: 500 })
    }
  }
}
