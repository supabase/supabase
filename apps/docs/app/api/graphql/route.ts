import { type DocumentNode, graphql, GraphQLError, parse, specifiedRules, validate } from 'graphql'
import { createComplexityLimitRule } from 'graphql-validation-complexity'
import { NextResponse } from 'next/server'
import { z } from 'zod'
import { ApiError, InvalidRequestError } from '~/app/api/utils'
import { rootGraphQLResolver } from '~/resources/rootResolver'
import { rootGraphQLSchema } from '~/resources/rootSchema'
import { createQueryDepthLimiter } from './validators'

export const runtime = 'edge'

const graphQLRequestSchema = z.object({
  query: z.string(),
  variables: z.record(z.any()).optional(),
  operationName: z.string().optional(),
})

async function handleGraphQLRequest(request: Request): Promise<NextResponse> {
  const body = await request.json().catch(() => {
    throw new InvalidRequestError('Request body must be valid JSON')
  })
  const parsedBody = graphQLRequestSchema.safeParse(body)
  if (!parsedBody.success) {
    const errorMessage = parsedBody.error?.message
    throw new InvalidRequestError(
      `Request body must be a valid GraphQL request object${errorMessage ? `: ${errorMessage}` : ''}`
    )
  }

  const { query, variables, operationName } = parsedBody.data
  const validationErrors = validateGraphQLRequest(query)
  if (validationErrors.length > 0) {
    return NextResponse.json({
      errors: validationErrors.map((error) => ({
        message: error.message,
        locations: error.locations,
        path: error.path,
      })),
    })
  }

  const result = await graphql({
    schema: rootGraphQLSchema,
    rootValue: rootGraphQLResolver,
    contextValue: { request },
    source: query,
    variableValues: variables,
    operationName,
  })
  // return NextResponse.json(result)

  // For now, just return a 404 response
  return new NextResponse('Not Found', { status: 404 })
}

function validateGraphQLRequest(query: string): ReadonlyArray<GraphQLError> {
  const MAX_DEPTH = 9

  let documentAST: DocumentNode
  try {
    documentAST = parse(query)
  } catch (error: unknown) {
    if (error instanceof GraphQLError) {
      return [error]
    } else {
      throw error
    }
  }
  const validationRules = [
    ...specifiedRules,
    createQueryDepthLimiter(MAX_DEPTH),
    createComplexityLimitRule(1500, {
      scalarCost: 1,
      objectCost: 2,
      listFactor: 10,
    }),
  ]
  return validate(rootGraphQLSchema, documentAST, validationRules)
}

export async function POST(request: Request): Promise<NextResponse> {
  try {
    return await handleGraphQLRequest(request)
  } catch (error: unknown) {
    console.error(error)

    if (error instanceof ApiError) {
      return new NextResponse(error.isPrivate() ? 'Internal Server Error' : error.message, {
        status: error.statusCode(),
      })
    } else {
      return new NextResponse('Internal Server Error', { status: 500 })
    }
  }
}
