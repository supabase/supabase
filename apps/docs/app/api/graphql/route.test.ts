import { beforeEach, describe, expect, it, vi } from 'vitest'
import { POST } from './route'

describe('/api/graphql basic error statuses', () => {
  beforeEach(() => {
    vi.clearAllMocks()
    vi.spyOn(console, 'error').mockImplementation(() => {})
  })

  it('should return error if request body is not valid JSON', async () => {
    const request = new Request('http://localhost/api/graphql', {
      method: 'POST',
      body: 'not json',
    })

    const response = await POST(request)
    const json = await response.json()
    expect(json.errors[0].message).toBe('Invalid request: Request body must be valid JSON')
  })

  it('should return error if request body is missing required fields', async () => {
    const request = new Request('http://localhost/api/graphql', {
      method: 'POST',
      body: JSON.stringify({ variables: {} }),
    })

    const response = await POST(request)
    const json = await response.json()
    expect(json.errors[0].message).toContain(
      'Invalid request: Request body must be valid GraphQL request object'
    )
  })

  it('should return error if query is not a string', async () => {
    const request = new Request('http://localhost/api/graphql', {
      method: 'POST',
      body: JSON.stringify({ query: 123 }),
    })

    const response = await POST(request)
    const json = await response.json()
    expect(json.errors[0].message).toContain(
      'Invalid request: Request body must be valid GraphQL request object'
    )
  })

  it('should return error for internal server errors', async () => {
    const request = new Request('http://localhost/api/graphql', {
      method: 'POST',
      body: JSON.stringify({ query: 'query { hello }' }),
    })

    vi.spyOn(request, 'json').mockImplementation(() => {
      throw new Error('Unexpected error')
    })

    const response = await POST(request)
    const json = await response.json()
    expect(json.errors[0].message).toBe('Internal Server Error')
  })
})

describe('/api/graphql schema snapshot', () => {
  it('should match snapshot', async () => {
    const schemaQuery = `
        query {
          schema
        }
      `
    const request = new Request('http://localhost/api/graphql', {
      method: 'POST',
      body: JSON.stringify({ query: schemaQuery }),
    })

    const response = await POST(request)
    const json = await response.json()
    expect(json.errors).toBeUndefined()

    const {
      data: { schema },
    } = json
    expect(schema).toMatchSnapshot()
  })
})
