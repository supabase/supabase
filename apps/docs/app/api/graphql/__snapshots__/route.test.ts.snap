// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`/api/graphql schema snapshot > should match snapshot 1`] = `
"schema {
  query: RootQueryType
}

"""
A document containing content from the Supabase docs. This is a guide, which might describe a concept, or explain the steps for using or implementing a feature.
"""
type Guide implements SearchResult {
  """The title of the document"""
  title: String

  """The URL of the document"""
  href: String

  """
  The full content of the document, including all subsections (both those matching and not matching any query string) and possibly more content
  """
  content: String

  """
  The subsections of the document. If the document is returned from a search match, only matching content chunks are returned. For the full content of the original document, use the content field in the parent Guide.
  """
  subsections: SubsectionCollection
}

"""Document that matches a search query"""
interface SearchResult {
  """The title of the matching result"""
  title: String

  """The URL of the matching result"""
  href: String

  """The full content of the matching result"""
  content: String
}

"""
A collection of content chunks from a larger document in the Supabase docs.
"""
type SubsectionCollection {
  """A list of edges containing nodes in this collection"""
  edges: [SubsectionEdge!]!

  """The nodes in this collection, directly accessible"""
  nodes: [Subsection!]!

  """The total count of items available in this collection"""
  totalCount: Int!
}

"""An edge in a collection of Subsections"""
type SubsectionEdge {
  """The Subsection at the end of the edge"""
  node: Subsection!
}

"""A content chunk taken from a larger document in the Supabase docs"""
type Subsection {
  """The title of the subsection"""
  title: String

  """The URL of the subsection"""
  href: String

  """The content of the subsection"""
  content: String
}

"""
A reference document containing a description of a function from a Supabase client library
"""
type ClientLibraryFunctionReference implements SearchResult {
  """The title of the document"""
  title: String

  """The URL of the document"""
  href: String

  """The content of the reference document, as text"""
  content: String

  """The programming language for which the function is written"""
  language: Language!

  """The name of the function or method"""
  methodName: String
}

enum Language {
  JAVASCRIPT
  SWIFT
  DART
  CSHARP
  KOTLIN
  PYTHON
}

type RootQueryType {
  """Get the GraphQL schema for this endpoint"""
  schema: String!

  """Search the Supabase docs for content matching a query string"""
  searchDocs(query: String!, limit: Int): SearchResultCollection
}

"""A collection of search results containing content from Supabase docs"""
type SearchResultCollection {
  """A list of edges containing nodes in this collection"""
  edges: [SearchResultEdge!]!

  """The nodes in this collection, directly accessible"""
  nodes: [SearchResult!]!

  """The total count of items available in this collection"""
  totalCount: Int!
}

"""An edge in a collection of SearchResults"""
type SearchResultEdge {
  """The SearchResult at the end of the edge"""
  node: SearchResult!
}"
`;
